File: 200-number-of-islands/200-number-of-islands.py
class Solution:
    def dfs(self, grid:List[List[str]], visited:List[List[bool]], x, y):
        m = len(grid)
        n = len(grid[0])
        if x < 0 or x >= m or y < 0 or y >= n or visited[x][y] or grid[x][y]=='0':
            return
        visited[x][y] = True
        self.dfs(grid, visited, x-1, y)
        self.dfs(grid, visited, x+1, y)
        self.dfs(grid, visited, x, y-1)
        self.dfs(grid, visited, x, y+1)
        
    def numIslands(self, grid: List[List[str]]) -> int:
        m = len(grid)
        n = len(grid[0])        
        visited = [[False]*n for i in range(m)]
        res = 0
        for i in range(m):
            for j in range(n):
                if visited[i][j] == False and grid[i][j] == '1':
                    res +=1
                    self.dfs(grid, visited, i, j)
                
        return res        
        

File: 200-number-of-islands/200-number-of-islands.java
class Solution {
    public int numIslands(char[][] grid) {
        int m = grid.length;
        int n = grid[0].length;
        boolean[][] visited = new boolean[m][n];
        int count = 0;
        for(int i=0;i<m;i++){
            for(int j=0;j<n;j++){
                if(!visited[i][j] && grid[i][j] == '1'){
                    count++;
                    dfs(grid, i, j, visited);
                }
            }
        }
        return count;
    }
    
    public void dfs(char[][] grid, int x, int y, boolean[][] visited){
        int m = grid.length;
        int n = grid[0].length;
        if(x < 0 || x >= m || y < 0 || y >= n || visited[x][y] || grid[x][y]=='0')
            return;
        visited[x][y] = true;
        dfs(grid, x+1, y, visited);
        dfs(grid, x-1, y, visited);
        dfs(grid, x, y+1, visited);
        dfs(grid, x, y-1, visited);
    }
}